File: backend/src/main/java/com/school/management/system/service/AdminService.java
package com.school.management.system.service;

import com.school.management.system.entity.ClassDetails;
import com.school.management.system.entity.Fees;
import com.school.management.system.entity.Student;
import com.school.management.system.entity.Teacher;
import com.school.management.system.payload.request.ClassRequestDTO;
import com.school.management.system.payload.request.FeeRequestDTO;
import com.school.management.system.payload.request.StudentRequestDTO;
import com.school.management.system.payload.request.StudentWithParentDTO;
import com.school.management.system.payload.request.TeacherRequestDTO;
import com.school.management.system.payload.response.DashboardStatsDTO;
import com.school.management.system.payload.response.FeeDetailsDTO;
import java.util.List;

public interface AdminService {

    DashboardStatsDTO getDashboardStats();

    // Student Methods
    Student createStudent(StudentWithParentDTO studentWithParentDTO);
    List<Student> getAllStudents();
    Student updateStudent(Long studentId, StudentRequestDTO studentDTO);
    void deleteStudent(Long studentId);

    // Teacher Methods
    List<Teacher> getAllTeachers();
    Teacher createTeacher(TeacherRequestDTO teacherDTO);
    Teacher updateTeacher(Long teacherId, TeacherRequestDTO teacherDTO);
    void deleteTeacher(Long teacherId);

    // Fee Methods
    List<FeeDetailsDTO> getAllFeeDetails();
    List<FeeDetailsDTO> getFeeDetailsByClass(Long classId);
    Fees addOrUpdateFee(FeeRequestDTO feeRequestDTO);

    // Class Methods
    List<ClassDetails> getAllClasses();
    ClassDetails createClass(ClassRequestDTO classDTO);
    ClassDetails updateClass(Long classId, ClassRequestDTO classDTO);
    void deleteClass(Long classId);
}

File: frontend/src/app/modules/admin/services/admin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_CONFIG } from '../../../core/config/api.config';

const API_URL = API_CONFIG.ADMIN + '/';

@Injectable({
  providedIn: 'root'
})
export class AdminService {

  constructor(private http: HttpClient) { }

  getDashboardStats(): Observable<any> {
    return this.http.get(API_URL + 'dashboard/stats');
  }

  getAllFees(): Observable<any> {
    return this.http.get(API_URL + 'fees');
  }

  sendFeeReminder(studentId: number): Observable<any> {
    return this.http.post(API_URL + `fees/reminder/${studentId}`, {});
  }

  getAllStudents(): Observable<any> {
    return this.http.get(API_URL + 'students');
  }

  createStudent(studentData: any): Observable<any> {
    return this.http.post(API_URL + 'students', studentData);
  }

  updateStudent(id: number, studentData: any): Observable<any> {
    return this.http.put(API_URL + `students/${id}`, studentData);
  }

  deleteStudent(id: number): Observable<any> {
    return this.http.delete(API_URL + `students/${id}`, { responseType: 'text' });
  }

  // --- Teacher Management ---
  getAllTeachers(): Observable<any> {
    return this.http.get(API_URL + 'teachers');
  }

  createTeacher(teacherData: any): Observable<any> {
    return this.http.post(API_URL + 'teachers', teacherData);
  }

  updateTeacher(id: number, teacherData: any): Observable<any> {
    return this.http.put(API_URL + `teachers/${id}`, teacherData);
  }

  deleteTeacher(id: number): Observable<any> {
    return this.http.delete(API_URL + `teachers/${id}`, { responseType: 'text' });
  }

  // --- Class Management ---
  getAllClasses(): Observable<any> {
    return this.http.get(API_URL + 'classes');
  }

  createClass(classData: any): Observable<any> {
    return this.http.post(API_URL + 'classes', classData);
  }

  updateClass(id: number, classData: any): Observable<any> {
    return this.http.put(API_URL + `classes/${id}`, classData);
  }

  deleteClass(id: number): Observable<any> {
    return this.http.delete(API_URL + `classes/${id}`, { responseType: 'text' });
  }

  // --- New Fee Management Methods ---
  getFeesByClass(classId: number): Observable<any> {
    return this.http.get(API_URL + `fees/class/${classId}`);
  }

  addOrUpdateFee(feeData: any): Observable<any> {
    return this.http.post(API_URL + 'fees', feeData);
  }

}
